Optim foreach cleanup for ASGD

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* __->__ #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to ASGD optimizer + cleanup

Differential Revision: [D33767867](https://our.internmc.facebook.com/intern/diff/D33767867)

mikaylagawarecki committed:
Optim foreach cleanup for ASGD

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/asgd.py:
```diff
@@ -76,19 +80,117 @@ def step(self, closure=None):
+    if weight_decay != 0:
+        torch._foreach_add_(grads, params, alpha=weight_decay)
+
+    # decay term
```
The initial `multitensor/_functional` ASGD logic assumes that eta is the same for all params passed to the function, (`eta = states[0]['eta']` where states[0] is the state for the first param) and then (`torch._foreach_mul_(params, 1 - lambd * eta)`) I wasn't sure whether this should be the case as it seemed slightly different from the other optimizers, so I made a change here, let me know if i should change it back!!

albanD replied:
Again I think it is safer to not change the behavior in this PR.
We could add a check that `all(states[0]['eta'] == states[i]['eta'] for i in range(len(states)))` if you think this is a bug. But I would prefer delay any change of behavior to a later PR to keep things simple in case we break things and we have to track back and revert some of these changes.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

albanD reviewed (commented):
Looks good. We should revert the change of behavior though.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

albanD reviewed (approved):
👍

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for ASGD"


Differential Revision: [D33767867](https://our.internmc.facebook.com/intern/diff/D33767867)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767867).
